### ISSUE TITLE
Some KEPs are not being picked by the KEP tool for annual reports

### ISSUE THREAD

**Issue by soltysh** [kubernetes/community#7213]
### Describe the issue

While working on [SIG Apps report](https://github.com/kubernetes/community/pull/7210) and [SIG CLI report](https://docs.google.com/document/d/1g3PCni-uVCSu4WdCaEbciQGky1J2Xm2GLSi-w56xq3A/edit?pli=1) I've noticed some of the KEPs not being picked by the tool. 

Specific examples for SIG-Apps are:
- 3140 - TimeZone support in CronJob
- 3017 - Pod Healthy Policy for PDB

for SIG-CLI:
- 3515 - Kubectl Explain OpenAPIv3
- 2227 - kubectl default container
- 1440 - Kubectl events

/cc @kubernetes/steering-committee @kubernetes/sig-contributor-experience

**Comment by Priyankasaggu11929** [comment]
Thanks for pointing @soltysh and @MadhavJivrajani. I'll take a look.

/assign

**Comment by palnabarun** [comment]
I tinkered a little on this issue over the weekend. There are two major issues contributing to the above:
1. The KEP metadata is inconsistent. Some have `latest-milestone` prefixed with `v`, and some don't. 
2. The logic for filtering the KEPs in the generator right now only considers `latest-milestone`. It should consider all the milestones of a KEP while determining if it was worked on in the target year/release.

**Comment by palnabarun** [comment]
Fixing the KEP metadata is out of scope for this repo, and I will take this up in k/enhancements. To unblock this issue, it can be considered that the `latest-milestone` isn't prefixed with `v`.

**Comment by Priyankasaggu11929** [comment]
Few observations:

The tool did not pick up the [listed KEPs](https://github.com/kubernetes/community/issues/7213#issue-1639493662) because (as of the time of writing this comment), it was trying to fetch KEPs with `latest-milestone` field set to either `"1.26"`, `"1.25"`, or `"1.24"` .

However, the latest-milestone for the listed KEPs had been updated to `"v1.27"`, which caused the tool to miss them.

(exception - "1440 - Kubectl events" - I'm still not sure why this was missed by the tool)

---

> The KEP metadata is inconsistent. Some have latest-milestone prefixed with v, and some don't.

The [current logic for filtering KEPs](https://github.com/kubernetes/community/blob/master/generator/app.go#L168-L173) only considers the suffixes of the `latest-milestone` field, so values with or without the "v" prefix are both considered valid, such as "v1.26" or "1.26", are both valid values currently.

---

>The logic for filtering the KEPs in the generator right now only considers latest-milestone. It should consider all the milestones of a KEP while determining if it was worked on in the target year/release.

As far as I understand, the `latest-milestone` field is the one used to determine whether a KEP is/was worked on in a particular release cycle(s).

The milestone fields (alpha/beta/stable) are usually filled out by the KEP owners in advance as part of their proactive plan for the KEP's implementation. So, this may or may-not always accurately reflect whether a milestone was achieved in a release cycle.

But I agree that `latest-milestone` field in KEPs is subject to change (to then active release cycle, if worked upon), so, the filtering logic would definitely require some updating.

@palnabarun, need more inputs from you to move ahead with it.

**Comment by palnabarun** [comment]
> The [current logic for filtering KEPs](https://github.com/kubernetes/community/blob/master/generator/app.go#L168-L173) only considers the suffixes of the latest-milestone field, so values with or without the "v" prefix are both considered valid, such as "v1.26" or "1.26", are both valid values currently.

The suffix being tested against is prefixed, i.e., the releases that you are fetching from GitHub Releases are prefixed with "v". Hence, the `strings.HasSuffix` check will return `false` for version strings not prefixed with "v". I could replicate this easily by changing the conditions a little.

You can try this:
```go
package main

import (
	"fmt"
	"strings"
)

func main() {
	latestMilestonePrefixed := "v1.27"
	latestMilestoneNotPrefixed := "1.27"

        releaseStringFromGitHub := "v1.27"

	fmt.Println(strings.HasSuffix(latestMilestonePrefixed, releaseStringFromGitHub))
	fmt.Println(strings.HasSuffix(latestMilestoneNotPrefixed, releaseStringFromGitHub))
}
```
---

> As far as I understand, the latest-milestone field is the one used to determine whether a KEP is/was worked on in a particular release cycle(s).

No. It is the latest release in which the KEP was worked on. This field doesn't capture history accurately.


```
# The most recent milestone for which work toward delivery of this KEP has been
# done. This can be the current (upcoming) milestone, if it is being actively
# worked on.
latest-milestone: "v1.19"
```

> The milestone fields (alpha/beta/stable) are usually filled out by the KEP owners in advance as part of their proactive plan for the KEP's implementation. So, this may or may-not always accurately reflect whether a milestone was achieved in a release cycle.

They can fill it out in advance, but they must update the state when milestones are completed. If the data is stale, that's a different problem. The annual reports then become the right time for them to go back and fix if there's a discrepancy in the data. See the docs for the field:

```
# The milestone at which this feature was, or is targeted to be, at each stage.
milestone:
  alpha: "v1.19"
  beta: "v1.20"
  stable: "v1.22"
```


The `milestone` field is supposed to be the actual state of the history of the KEP. If a certain milestone is in feature, our logic would naturally ignore it.

**Comment by palnabarun** [comment]
I've raised https://github.com/kubernetes/community/pull/7227, which tries to fix this by checking across all `milestone`s defined for a KEP.

**Comment by Priyankasaggu11929** [comment]
@palnabarun, Thanks for the clarification in https://github.com/kubernetes/community/issues/7213#issuecomment-1488356961 & the PR. :)
